geomstats/geometry/product_manifold.py

```diff
@@ -311,3 +312,263 @@ def is_tangent(self, vector, base_point, atol=gs.atol):
+    ----------
+    base_metric : RiemannianMetric
+        Base metric.
+    n_copies : int
+        Number of replication of the base metric.
+    base_shape : tuple of int
+        Shape of one element of the underlying base manifold
+    """
+
+    def __init__(self, base_metric, n_copies, base_shape):
+        geomstats.errors.check_integer(n_copies, "n_copies")
+        dim = n_copies * base_metric.dim
+        super(NFoldMetric, self).__init__(dim=dim)
+        self.base_shape = base_shape
+        self.base_metric = base_metric
+        self.n_copies = n_copies
+
+    def metric_matrix(self, base_point=None):
+        """Compute the matrix of the inner-product.
+
+        Matrix of the inner-product defined by the Riemmanian metric
+        at point base_point of the manifold.
+
+        Parameters
+        ----------
+        base_point : array-like, shape=[..., n_copies, base_shape]
+            Point on the manifold at which to compute the inner-product matrix.
+            Optional, default: None.
+
+        Returns
+        -------
+        matrix : array-like, shape=[..., n_copies, dim, dim]
+            Matrix of the inner-product at the base point.
+        """
+        point_ = gs.reshape(base_point, (-1,) + self.base_shape)
+        matrices = self.base_metric.metric_matrix(point_)
+        dim = self.base_metric.dim
+        reshaped = gs.reshape(matrices, (-1, self.n_copies, dim, dim))
+        return gs.squeeze(reshaped)
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point):
+        """Compute the inner-product of two tangent vectors at a base point.
+
+        Inner product defined by the Riemannian metric at point `base_point`
+        between tangent vectors `tangent_vec_a` and `tangent_vec_b`.
+
+        Parameters
+        ----------
+        tangent_vec_a : array-like, shape=[..., n_copies, base_shape]
+            First tangent vector at base point.
+        tangent_vec_b : array-like, shape=[..., n_copies, base_shape]
+            Second tangent vector at base point.
+        base_point : array-like, shape=[..., n_copies, base_shape]
+            Point on the manifold.
+            Optional, default: None.
+
+        Returns
+        -------
+        inner_prod : array-like, shape=[...,]
+            Inner-product of the two tangent vectors.
+        """
+        point_ = gs.reshape(base_point, (-1,) + self.base_shape)
+        vector_a = gs.reshape(tangent_vec_a, (-1,) + self.base_shape)
+        vector_b = gs.reshape(tangent_vec_b, (-1,) + self.base_shape)
+        inner_each = self.base_metric.inner_product(vector_a, vector_b, point_)
+        reshaped = gs.reshape(inner_each, (-1, self.n_copies))
+        return gs.squeeze(gs.sum(reshaped, axis=-1))
+
+    def exp(self, tangent_vec, base_point, **kwargs):
+        """Compute the Riemannian exponential of a tangent vector.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., n_copies, base_shape]
```

Comment:
technically, it would be *base_shape, but maybe we should keep it this way otherwise it's going to be hard to read.